### PR TITLE
feat(sql) Support dense_rank() function

### PR DESCRIPTION
**Pull request by hennessydylan**
from
https://github.com/questdb/questdb/issues/2578

need some guidance

### CODE REVIEW

File: core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java [hunk 1]
```diff
@@ -1776,6 +1776,387 @@ public void testRowNumberWithPartitionAndOrderBySymbolWildcardLast() throws Exce
...
+                        " rnd_symbol('AA','BB','CC') symbol," +
+                        " timestamp_sequence(0, 100000000000) ts" +
+                        " from long_sequence(10)" +
+                        ") timestamp(ts) partition by day",
+                null,
+                true,
+                false
+        );
+    }
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1970]: I don't think testing with wildcard before and after window function is necessary .

File: core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java [hunk 2]
```diff
@@ -1776,6 +1776,387 @@ public void testRowNumberWithPartitionAndOrderBySymbolWildcardLast() throws Exce
...
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n" +
+                        "1\n",
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1924]: Typo -> dennse_rank

File: core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java [hunk 3]
```diff
@@ -1776,6 +1776,387 @@ public void testRowNumberWithPartitionAndOrderBySymbolWildcardLast() throws Exce
         );
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1778]: Cases that'd be good to test :
- OVER (ORDER BY ts)
- OVER  (PARTITION BY something) (no ORDER BY)
- PARTITION BY more than one column 
- ORDER BY more than one column

File: core/src/main/java/io/questdb/griffin/engine/functions/window/DenseRankFunctionFactory.java [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+import io.questdb.griffin.engine.orderby.RecordComparatorCompiler;
+import io.questdb.griffin.engine.window.WindowContext;
+import io.questdb.griffin.engine.window.WindowFunction;
+import io.questdb.std.IntList;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
+import io.questdb.std.Unsafe;
+
+public class DenseRankFunctionFactory implements FunctionFactory {
+
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/DenseRankFunctionFactory.java:49]: Should be "dense_rank()" .

File: core/src/main/java/io/questdb/griffin/engine/functions/window/DenseRankFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+            map.clear();
+        }
+    }
+
+    private static class SequenceDenseRankFunction extends LongFunction implements ScalarFunction, WindowFunction, Reopenable {
+
+        private int columnIndex;
+
+        private long rank;
+
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/DenseRankFunctionFactory.java:313]: Constructor should match class name:  SequenceRankFunction -> SequenceDenseRankFunction .

File: core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java [hunk 6]
```diff
@@ -1776,6 +1776,387 @@ public void testRowNumberWithPartitionAndOrderBySymbolWildcardLast() throws Exce
...
 
+/* dense_rank() function tests
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1779]: These comments don't seem necessary or important .

File: core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java (hunk 1 above)
Comment (hennessydylan) [reply core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1970]: Agreed didn't realize that this was redundant

Comment (hennessydylan) [reply core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionTest.java:1970]: Removed Wildcard First & Last from function test names to reduce redundancy

**Comment by hennessydylan** [comment]
Re trying this

**Pull request closed without merging**